### PR TITLE
generated texts from the plugin settings and passed the same to gener…

### PR DESCRIPTION
**Pull request by pradeepmurugesan**
…ate index

#### Summary
Add the plugin text to the index, so that the plugin settings are searchable in the system console

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/11897

#### Related Pull Requests
- Has server changes - No
- Has redux changes - No
- Has mobile changes - No

### CODE REVIEW

**Comment by hanzei** [comment]
@pradeepmurugesan Let me know when this is ready for review

**Comment by hanzei** [comment]
@aaronrothschild Gentle reminder to review this PR

File: components/admin_console/admin_sidebar/admin_sidebar.jsx [hunk 1]
```diff
@@ -88,6 +89,26 @@ export default class AdminSidebar extends React.Component {
         }
 
         this.updateTitle();
+
+        if (!Utils.isMobile()) {
+            $('.admin-sidebar .nav-pills__container').perfectScrollbar({
+                suppressScrollX: true,
```
Comment (marianunez) [inline components/admin_console/admin_sidebar/admin_sidebar.jsx]: This change does not seem related to this PR, is this needed?

File: components/admin_console/admin_sidebar/admin_sidebar.jsx [hunk 2]
```diff
@@ -88,6 +89,26 @@ export default class AdminSidebar extends React.Component {
...
+    componentDidUpdate(prevProps) {
+        if (this.idx != null &&
+            (!isEqual(this.props.plugins, prevProps.plugins) ||
+                !isEqual(this.props.adminDefinition, prevProps.adminDefinition))) {
+            this.idx = generateIndex(this.props.adminDefinition, this.context.intl, this.props.plugins);
+        }
+
+        if (!Utils.isMobile()) {
+            $('.admin-sidebar .nav-pills__container').perfectScrollbar({
+                suppressScrollX: true,
```
Comment (marianunez) [inline components/admin_console/admin_sidebar/admin_sidebar.jsx]: Same comment as above regarding this scrollbar change

**Comment by prapti** [comment]
@ali-farooq0 Gave this a quick check and looks like the search has issues similar (or same) as the one you noted in https://github.com/mattermost/mattermost-webapp/pull/3865#issue-324330090? For example, in the screenshot, searching for "git" shows plugins that should not show up, and searching for "Enable Plugin" does not show all plugins. Thoughts? 

In the screens
Screen Shot 2019-10-08 at 5 07 45 PM
Screen Shot 2019-10-08 at 5 07 28 PM

File: components/admin_console/admin_sidebar/admin_sidebar.jsx [hunk 3]
```diff
@@ -262,6 +283,11 @@ export default class AdminSidebar extends React.Component {
                     }
                 }
 
+                if (this.state.sections != null) {
+                    if (this.state.sections.indexOf(`plugin_${p.id}`) === -1) {
```
Comment (marianunez) [inline components/admin_console/admin_sidebar/admin_sidebar.jsx]: Is this check necessary here? It seems that the URL check is already covered [here](https://github.com/mattermost/mattermost-webapp/blob/20142963730312102b5084035ca24bdb613fad22/components/admin_console/admin_sidebar/admin_sidebar.jsx#L204) in the `renderRootMenu`

**Comment by marianunez** [comment]
> For example, in the screenshot, searching for "git" shows plugins that should not show up, and searching for "Enable Plugin" does not show all plugins. Thoughts?

@prapti I believe that in this case is different because it searches not only the plugin name that you see directly in the sections result list but also all the settings (name, help text and key) within that section. 

Could you verify on your test that the sections that do not have `Git` highlighted on the name, you can click on them and see in the settings page somewhere something containing that search term?

File: components/admin_console/admin_sidebar/admin_sidebar.jsx (hunk 3 above)
Comment (pradeepmurugesan) [reply components/admin_console/admin_sidebar/admin_sidebar.jsx]: Yes.. The root menu iterates over the static admin definitions. This is required when iterating the available pluings and render the matched ones.

File: components/admin_console/admin_sidebar/admin_sidebar.jsx (hunk 2 above)
Comment (pradeepmurugesan) [reply components/admin_console/admin_sidebar/admin_sidebar.jsx]: Not sure how these came in..Thanks for spotting. Removed 👍

File: components/admin_console/admin_sidebar/admin_sidebar.jsx (hunk 1 above)
Comment (pradeepmurugesan) [reply components/admin_console/admin_sidebar/admin_sidebar.jsx]: Not sure how these came in..Thanks for spotting. Removed 👍

**Comment by pradeepmurugesan** [comment]
> > For example, in the screenshot, searching for "git" shows plugins that should not show up, and searching for "Enable Plugin" does not show all plugins. Thoughts?
> 
> @prapti I believe that in this case is different because it searches not only the plugin name that you see directly in the sections result list but also all the settings (name, help text and key) within that section.
> 
> Could you verify on your test that the sections that do not have `Git` highlighted on the name, you can click on them and see in the settings page somewhere something containing that search term?

Yes.. @prapti . The search considers the name, settings and few other attributes of plugins and filters them. Kindly go into the filtered sections and see if it contains the search text

File: components/admin_console/admin_sidebar/admin_sidebar.jsx (hunk 3 above)
Comment (marianunez) [reply components/admin_console/admin_sidebar/admin_sidebar.jsx]: Gotcha!

**Comment by prapti** [comment]
@marianunez @pradeepmurugesan I did look for the search term in the settings for the plugins  (setting options, help text etc). For example in the following screenshot, notice that Antivirus plugin shows up on the search result for "git" and nowhere within Antivirus there is "git" (ran a quick cmd+f browser search as well). Same is the case for Autolink, Custom User Attributes, Jenkins and Welcome bot.

The other thing to notice is that not all plugins come up in the result when "Enable Plugin" is searched although there is an Enable Plugin setting for every plugin.

**Comment by marianunez** [comment]
> @marianunez @pradeepmurugesan I did look for the search term in the settings for the plugins (setting options, help text etc). For example in the following screenshot, notice that Antivirus plugin shows up on the search result for "git" and nowhere within Antivirus there is "git" (ran a quick cmd+f browser search as well). Same is the case for Autolink, Custom User Attributes, Jenkins and Welcome bot.
> 

Thanks for testing this more @prapti. I believe the reason those plugins are coming up as a result is do to the markdown that contains links which point to a github url and causing a match on that. @pradeepmurugesan we would need to have a way to handle the markdown and ignore the actual URLs. 

> The other thing to notice is that not all plugins come up in the result when "Enable Plugin" is searched although there is an Enable Plugin setting for every plugin.
> 

"Enable Plugin" is a special setting given that is not part of the plugin's manifest but is always included in the settings screen for each plugin. I'm a bit surprised that is currently not showing given that it currently works properly before this PR.

File: utils/admin_console_plugin_index.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+function extractTextsFromPlugin(plugin) {
```
Comment (ali-farooq0) [inline utils/admin_console_plugin_index.js]: missing a line here.

File: utils/admin_console_plugin_index.test.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import {samplePlugin1, samplePlugin2} from '../tests/helpers/admin_console_plugin_index_sample_pluings';
```
Comment (ali-farooq0) [inline utils/admin_console_plugin_index.test.js]: missing a line between import and the license description

File: utils/admin_console_plugin_index.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            }
+        }
+    }
+    return texts;
+}
+
+export function getPluginEntries(plugins = {}) {
+    const entries = {};
+    for (const pluginId of Object.keys(plugins || {})) {
+        const url = `plugin_${pluginId}`;
```
Comment (ali-farooq0) [inline utils/admin_console_plugin_index.js]: Is this actually a url? i.e. this can be hyperlinked?

File: utils/admin_console_plugin_index.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                texts.push(setting.key);
+            }
+        }
+    }
+    return texts;
+}
+
+export function getPluginEntries(plugins = {}) {
+    const entries = {};
+    for (const pluginId of Object.keys(plugins || {})) {
```
Comment (ali-farooq0) [inline utils/admin_console_plugin_index.js]: lets move the initialization of plugins at the start of the function, that way we'll have one spot where we do all the checking.

```
export function getPluginEntries(pluginsObj) {
const entries = {};
const plugins = pluginsObj || {};
    for (const pluginId of Object.keys(plugins)) {

```

File: utils/admin_console_plugin_index.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    if (plugin.settings_schema) {
+        if (plugin.settings_schema.footer) {
+            texts.push(plugin.settings_schema.footer);
+        }
+        if (plugin.settings_schema.header) {
+            texts.push(plugin.settings_schema.header);
+        }
+    }
+
+    if (plugin.settings_schema && plugin.settings_schema.settings) {
```
Comment (ali-farooq0) [inline utils/admin_console_plugin_index.js]: can the `plugin.settings_schema.settings` check be moved into the if block on line 11? that way the parent check is in one spot?

File: components/admin_console/admin_sidebar/admin_sidebar.test.jsx [hunk 9]
```diff
@@ -5,8 +5,13 @@ import React from 'react';
 import {IntlProvider} from 'react-intl';
 import {shallow} from 'enzyme';
 
+import {samplePlugin1} from '../../../tests/helpers/admin_console_plugin_index_sample_pluings';
+
 import AdminSidebar from 'components/admin_console/admin_sidebar/admin_sidebar.jsx';
 import AdminDefinition from 'components/admin_console/admin_definition';
+import {generateIndex} from '../../../utils/admin_console_index';
```
Comment (ali-farooq0) [inline components/admin_console/admin_sidebar/admin_sidebar.test.jsx]: i think you should just be able to do `utils/admin_console_index';`

File: components/admin_console/admin_sidebar/admin_sidebar.test.jsx [hunk 10]
```diff
@@ -5,8 +5,13 @@ import React from 'react';
 import {IntlProvider} from 'react-intl';
 import {shallow} from 'enzyme';
 
+import {samplePlugin1} from '../../../tests/helpers/admin_console_plugin_index_sample_pluings';
```
Comment (ali-farooq0) [inline components/admin_console/admin_sidebar/admin_sidebar.test.jsx]: same with this `tests/helpers/admin_console_plugin_index_sample_pluings'` should suffice i think 🤔

File: components/admin_console/admin_sidebar/admin_sidebar.test.jsx [hunk 11]
```diff
@@ -5,8 +5,13 @@ import React from 'react';
...
+
 import AdminSidebar from 'components/admin_console/admin_sidebar/admin_sidebar.jsx';
 import AdminDefinition from 'components/admin_console/admin_definition';
+import {generateIndex} from '../../../utils/admin_console_index';
+
+const enMessages = require('../../../i18n/en');
```
Comment (ali-farooq0) [inline components/admin_console/admin_sidebar/admin_sidebar.test.jsx]: does this need to be `require`, can this just be `import en from 'i18n/en.json';`?

File: components/admin_console/admin_sidebar/admin_sidebar.test.jsx [hunk 12]
```diff
@@ -240,4 +247,142 @@ describe('components/AdminSidebar', () => {
...
+                'mattermost-autolink': samplePlugin1,
+            }});
+            expect(generateIndex).toHaveBeenCalledTimes(0);
+
+            wrapper.setProps({adminDefinition: AdminDefinition});
+            expect(generateIndex).toHaveBeenCalledTimes(0);
+        });
+    });
+
+    describe('Plugins', () => {
```
Comment (ali-farooq0) [inline components/admin_console/admin_sidebar/admin_sidebar.test.jsx]: nice tests 👍 !

File: tests/helpers/admin_console_plugin_index_sample_pluings.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+export const samplePlugin1 = {
+    id: 'mattermost-autolink',
+    name: 'Autolink',
+    description: 'Automatically rewrite text matching a regular expression into a Markdown link.',
+    version: '1.1.0',
+    settings_schema: {
+        header: 'Configure this plugin directly in the config.json file. Learn more [in our documentation](https://github.com/mattermost/mattermost-plugin-autolink/blob/master/README.md).\n\nTo report an issue, make a suggestion or a contribution, [check the plugin repository](https://github.com/mattermost/mattermost-plugin-autolink).',
+        footer: '',
```
Comment (ali-farooq0) [inline tests/helpers/admin_console_plugin_index_sample_pluings.js]: can we have a non empty footer value here, to include all the fields for testing.

File: tests/helpers/admin_console_plugin_index_sample_pluings.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        settings: [
+            {
+                key: 'EnableAdminCommand',
```
Comment (ali-farooq0) [inline tests/helpers/admin_console_plugin_index_sample_pluings.js]: and to include `label` key here, as we're using it to extract out text.

File: utils/admin_console_index.test.jsx [hunk 15]
```diff
@@ -5,6 +5,8 @@ import {IntlProvider} from 'react-intl';
 
 import AdminDefinition from 'components/admin_console/admin_definition.jsx';
 
+import {samplePlugin1, samplePlugin2} from '../tests/helpers/admin_console_plugin_index_sample_pluings';
```
Comment (ali-farooq0) [inline utils/admin_console_index.test.jsx]: `'tests/helpers/admin_console_plugin_index_sample_pluings'`

File: utils/admin_console_index.jsx [hunk 16]
```diff
@@ -75,16 +77,23 @@ export function adminDefinitionsToUrlsAndTexts(adminDefinition, intl) {
     return entries;
 }
 
-export function generateIndex(adminDefinition, intl) {
+export function generateIndex(AdminDefinition, intl, plugins) {
```
Comment (ali-farooq0) [inline utils/admin_console_index.jsx]: err feels a bit awkward to not have two similar definitions not close to each other, can we change this to `generateIndex(adminDefinition, plugins, intl)` that way the definitions are close to each other.

File: utils/admin_console_index.test.jsx [hunk 17]
```diff
@@ -72,4 +74,14 @@ describe('AdminConsoleIndex.generateIndex', () => {
         expect(idx.search('characters')).toEqual([]);
         expect(idx.search('notexistingword')).toEqual([]);
     });
+
+    it('should generate a index including the plugin settings', () => {
```
Comment (ali-farooq0) [inline utils/admin_console_index.test.jsx]: nice test 👍

File: utils/admin_console_plugin_index.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        if (plugin.settings_schema.settings) {
+            for (const setting of Object.values(plugin.settings_schema.settings)) {
+                if (setting.label) {
+                    texts.push(setting.label);
+                }
+                if (setting.display_name) {
+                    texts.push(setting.display_name);
+                }
+                if (setting.help_text) {
+                    texts.push(setting.help_text);
```
Comment (marianunez) [inline utils/admin_console_plugin_index.js]: The `help_text` contains markdown that can include urls that are not displayed and show false results as experienced in the preliminary testing. We need to strip this out before adding to the searchable texts, you can probably use [this util markdown](https://github.com/mattermost/mattermost-webapp/blob/c7493e03dbe40e56a1e4a7e6ca7138ea3d9284df/utils/markdown/index.js#L29) `stripMarkdown` function

File: utils/admin_console_plugin_index.js [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    }
+    if (plugin.id) {
+        texts.push(plugin.id);
+    }
+    if (plugin.settings_schema) {
+        if (plugin.settings_schema.footer) {
+            texts.push(plugin.settings_schema.footer);
+        }
+        if (plugin.settings_schema.header) {
+            texts.push(plugin.settings_schema.header);
```
Comment (marianunez) [inline utils/admin_console_plugin_index.js]: For `footer` and `header` applies same comment below regarding `help_text` markdown.

File: utils/admin_console_plugin_index.js (hunk 18 above)
Comment (pradeepmurugesan) [reply utils/admin_console_plugin_index.js]: done 👍

File: utils/admin_console_plugin_index.js (hunk 19 above)
Comment (pradeepmurugesan) [reply utils/admin_console_plugin_index.js]: done 👍

**Comment by pradeepmurugesan** [comment]
@ali-farooq0 @marianunez addressed the comments.. Kindly review 🙇

File: utils/admin_console_plugin_index.js (hunk 19 above)
Comment (marianunez) [reply utils/admin_console_plugin_index.js]: I still don't see the change of the markdown for `footer` or `header`. Maybe you missed to push these changes?

File: utils/admin_console_plugin_index.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        if (plugin.settings_schema.settings) {
+            for (const setting of Object.values(plugin.settings_schema.settings)) {
+                if (setting.label) {
+                    texts.push(stripMarkdown(setting.label));
```
Comment (marianunez) [inline utils/admin_console_plugin_index.js]: For `label` and `display_name` I don't believe it's necessary to strip markdown given that it is not supported in these fields.

**Comment by pradeepmurugesan** [comment]
@marianunez My Bad.. fixed the field now.

File: components/admin_console/admin_sidebar/admin_sidebar.test.jsx [hunk 21]
```diff
@@ -5,8 +5,13 @@ import React from 'react';
 import {IntlProvider} from 'react-intl';
 import {shallow} from 'enzyme';
 
+import {samplePlugin1} from 'tests/helpers/admin_console_plugin_index_sample_pluings';
+
 import AdminSidebar from 'components/admin_console/admin_sidebar/admin_sidebar.jsx';
 import AdminDefinition from 'components/admin_console/admin_definition';
+import {generateIndex} from 'utils/admin_console_index';
+
+const enMessages = require('i18n/en');
```
Comment (ali-farooq0) [inline components/admin_console/admin_sidebar/admin_sidebar.test.jsx]: minor nit: Should use es6 style import `import en from 'i18n/en';` here for consistency.

**Comment by ali-farooq0** [comment]
@prapti this is ready for testing 👍

**Comment by ali-farooq0** [comment]
orr actually @pradeepmurugesan do you mind merging `master` into this branch before @prapti starts testing.

**Comment by prapti** [comment]
> orr actually @pradeepmurugesan do you mind merging `master` into this branch before @prapti starts testing.

Did you want me to test post merge @ali-farooq0? I don't mind doing a quick review before merging.

**Comment by marianunez** [comment]
@prapti if you can do a quick test before merging so we can make sure your initial findings were addressed.

**Comment by prapti** [comment]
@marianunez @ali-farooq0 The search results look much cleaner now. Except I have one question that I had raised earlier too - searching for "Enable Plugin" does not show all the plugins. Given that all plugins have that field within their settings, shouldn't we expect all plugins to be listed when searched for it?

**Comment by pradeepmurugesan** [comment]
@marianunez I am sorry.. Subconsciously  rebased instead of merge.. 🤦‍♂

**Comment by pradeepmurugesan** [comment]
> @marianunez @ali-farooq0 The search results look much cleaner now. Except I have one question that I had raised earlier too - searching for "Enable Plugin" does not show all the plugins. Given that all plugins have that field within their settings, shouldn't we expect all plugins to be listed when searched for it?

@prapti looks like `Enable plugin` is not a setting mentioned explicitly in the plugin.json . Ex [github 
 plugin setting](https://github.com/mattermost/mattermost-plugin-github/blob/master/plugin.json)

Few plugin settings have other setting with Enable keyword. For ex  in the above plugin `EnablePrivateRepo` setting is present. That's y its filtered..

**Comment by prapti** [comment]
Thank you so much for clearing that confusion @pradeepmurugesan! That makes it a lot more easier to understand why that was happening :)

**Comment by marianunez** [comment]
> @prapti looks like `Enable plugin` is not a setting mentioned explicitly in the plugin.json . Ex [github plugin setting](https://github.com/mattermost/mattermost-plugin-github/blob/master/plugin.json)

@pradeepmurugesan This is correct that is not a setting of a plugin. However, the `Enable Plugin` setting is added to all plugin settings screen [here](https://github.com/mattermost/mattermost-webapp/blob/4b51ebb99a0038028881e2432c081f64ac0a41f7/components/admin_console/custom_plugin_settings/index.js#L41) and before this PR it did bring up all the plugins when searching for "Enable Plugin". 

Sorry to keep delaying this PR, but given that it is introducing a regression, could you take a look at what has changed that is no longer taking this setting into account for the search?

**Comment by prapti** [comment]
> A bit more investigation needs to be done for the "Enable Plugin" case - see comment above

Thanks for noting that Maria! I can check back after dev review since you've asked for changes.

**Comment by pradeepmurugesan** [comment]
@marianunez No problem regarding the delay. 👍 

BTW it's not a regression..  Previously no plugins were filtered. Only The plugin management section is indexed. If the search string matches any of the `Plugin Management` then all the plugins will be rendered. For ex: Let's try the search string `Marketplace`

Screenshot 2019-10-23 at 08 48 18
Screenshot 2019-10-23 at 08 48 27


Even thought the string `Marketplace` is not present in many plugins they will be shown. The logic is if the Parent ( Plugin management ) contains the search text then all the plugins will be shown. 👍 

Having said that I will look into why the custom setting is not indexed..

**Comment by pradeepmurugesan** [comment]
@marianunez I found the reason for `Enable Plugin` setting not being indexed. We are using the `props.plugins` to generate the index. `props.plugin` won't have the `Enable plugin` setting. 

The `Enable plugin` setting is added in `makeGetPluginSchema`. This was proposed as the initial solution and later @jespino and Myself decided to go with [props.plugin](https://community.mattermost.com/core/pl/1t6xg8wnyfbk3bsmhxuopwhqie).

Do we really need to have this setting indexed ? If yes, then we need to change the whole implementation.

**Comment by marianunez** [comment]
> @marianunez I found the reason for `Enable Plugin` setting not being indexed. We are using the `props.plugins` to generate the index. `props.plugin` won't have the `Enable plugin` setting.

Thanks for investigating this further! @pradeepmurugesan 

> Do we really need to have this setting indexed ? If yes, then we need to change the whole implementation.

For consistency I believe we should have this setting indexed. However, given that this is a special setting that is always added to the plugins setting, I propose a workaround where we always add this setting to the search index for each plugin in [extractTextsFromPlugin](https://github.com/mattermost/mattermost-webapp/blob/c96108b662ff8e231d1789a808342a8488b20c12/utils/admin_console_plugin_index.js#L6) as a special case. 

Thoughts? @pradeepmurugesan @jespino

**Comment by pradeepmurugesan** [comment]
> For consistency I believe we should have this setting indexed. However, given that this is a special setting that is always added to the plugins setting, I propose a workaround where we always add this setting to the search index for each plugin in [extractTextsFromPlugin](https://github.com/mattermost/mattermost-webapp/blob/c96108b662ff8e231d1789a808342a8488b20c12/utils/admin_console_plugin_index.js#L6) as a special case.
> 
> Thoughts? @pradeepmurugesan @jespino

@marianunez I still doubt if this is a valid business use case. I may be wrong, but since this setting is present in all the plugins and added by default, I am not sure how useful it will be when filtering.

I see the following cons when the `Enable plugin` setting is indexed.
 
1. If I type `enable plugin` in the search box, I can see all the plugins. Its not filtering any specific plugin

2. If a plugin has a setting starting with `enableXXXX` for ex `EnablePrivateRepo` in case of github plugin, the admin has to type the entire setting name to filter, as the keyword `enable` will list all the plugins.

If we can get some input from any Product person, it would be very helpful.

**Comment by marianunez** [comment]
> @marianunez I still doubt if this is a valid business use case. I may be wrong, but since this setting is present in all the plugins and added by default, I am not sure how useful it will be when filtering.
> 
> I see the following cons when the `Enable plugin` setting is indexed.
> 
> 1. If I type `enable plugin` in the search box, I can see all the plugins. Its not filtering any specific plugin

If it's a setting they can see in the screen, I believe this is the correct behavior.

> 2. If a plugin has a setting starting with `enableXXXX` for ex `EnablePrivateRepo` in case of github plugin, the admin has to type the entire setting name to filter, as the keyword `enable` will list all the plugins.
> 

Also, would be a behavior expected when doing a search of a term that is shared with other settings. Not showing would not be expected given that for the Administrator there is no diference between the `Enable Plugin` setting and others but that one does not show in results.

> If we can get some input from any Product person, it would be very helpful.

@aaronrothschild @jasonblais Thoughts on having all the plugins show as a result when the Admin searches `Enable Plugin` given that is a setting that is at the top of all plugin settings screen? See thread above

Screen Shot 2019-10-23 at 12 51 34 PM

**Comment by jasonblais** [comment]
@pradeepmurugesan @marianunez @prapti My proposal is to accept the changes as-is, and creating a low priority bug or story ticket for the known issue where `Enable Plugin` generated setting isn't searchable.

Why? Here are my brief thoughts:

For consistency, searching the term `Enable Plugin` should show each plugin in the sidebar since the setting is contained in each plugin setting page. I agree with that assessment.

However, this is an edge case, where for me the benefit to fix the edge case doesn't justify the cost of re-doing the entire implementation. Moreover, the changes here are already a good improvement to finding plugins in the System Console (e.g. searching for `jira` or `jenkins` now displays the plugin setting page in the sidebar whereas previously it would not), and does not introduce a regression to existing behaviour.

So with that, my proposal would be merging the changes as-is, and logging a low priority bug or story ticket for the known issue where `Enable Plugin` generated setting isn't searchable.

I'd be happy to discuss further!

cc @aaronrothschild

**Comment by marianunez** [comment]
> However, this is an edge case, where for me the benefit to fix the edge case doesn't justify the cost of re-doing the entire implementation.

@jasonblais Agreed on that it does not make sense redoing the entire implementation. I have proposed a simple workaround that would only include a couple of more lines in this PR. 

0/5 if it was to be deferred to a bug and the minor fix can be introduced in a separate PR.

**Comment by jasonblais** [comment]
> I have proposed a simple workaround above that would only include a couple of more lines in this PR.

@marianunez thank you for clarifying, I missed this when reading the above thread. If it's a minor fix, @pradeepmurugesan would you be open to helping with it?

For consistency, searching the term `Enable Plugin` should show each plugin in the sidebar since the setting is contained in each plugin setting page. While it is true that `enable` would return all pages, that is true for all other settings with an `Enable XXX` setting, which there are several.

**Comment by pradeepmurugesan** [comment]
> > I have proposed a simple workaround above that would only include a couple of more lines in this PR.
> 
> @marianunez thank you for clarifying, I missed this when reading the above thread. If it's a minor fix, @pradeepmurugesan would you be open to helping with it?

@jasonblais  I would be happy to. Just wanted to confirm if the behaviour is fine. Thanks for clarifying. 👍 

Thanks @marianunez. Will include the `Enable Plugin` setting for indexing

**Comment by pradeepmurugesan** [comment]
added the `Enable plugin` setting for indexing. 

Screenshot 2019-10-24 at 21 49 21


@prapti kindly test the latest server. @marianunez Kindly review my changes.. 🙇

File: components/admin_console/custom_plugin_settings/index.js [hunk 22]
```diff
@@ -6,12 +6,11 @@ import {createSelector} from 'reselect';
 
 import {getRoles} from 'mattermost-redux/selectors/entities/roles';
 
-import {Constants} from 'utils/constants';
-import {t} from 'utils/i18n';
 import SchemaAdminSettings from '../schema_admin_settings';
 import {it} from '../admin_definition';
 
 import CustomPluginSettings from './custom_plugin_settings.jsx';
+import getEnablePluginSetting from './EnablePluginSetting';
```
[the remark this answers is not in the document]
Comment (pradeepmurugesan) [reply components/admin_console/custom_plugin_settings/index.js]: done 👍

**Comment by prapti** [comment]
@pradeepmurugesan I hate to keep stalling this PR but looks like two of the plugins with Enable Plugin -- Custom User Attributes and Welcome Bot -- still don't show up in the search result (referring to tests done at the PR test server at https://mattermost-webapp-pr-3619.test.mattermost.cloud).
@marianunez any idea?

**Comment by pradeepmurugesan** [comment]
@prapti Thanks for testing. Not sure how I missed this in my previous fix. It's another corner case. Will fix the same and update the PR.

**Comment by pradeepmurugesan** [comment]
@prapti  Kindly test now.. 

@marianunez The reason why `Enable plugin` setting was not indexed was because the plugins has no settings and the code for indexing the `Enable plugin` setting was not reached.

Now indexing the default setting takes place first , followed by the `settings_schema` in the plugin.json. Kindly review
